gateway/gateway.go

```diff
@@ -667,14 +689,14 @@ func (gateway *HandleT) userWebRequestWorkerProcess(userWebRequestWorker *userWe
 		// update stats request wise
 		gateway.updateSourceStats(sourceStats, "gateway.write_key_requests", sourceTagMap)
 		gateway.updateSourceStats(sourceSuccessStats, "gateway.write_key_successful_requests", sourceTagMap)
-		gateway.updateSourceStats(sourceFailStats, "gateway.write_key_failed_requests", sourceTagMap)
+		gateway.updateFailedSourceStats(sourceFailStats, "gateway.write_key_failed_requests", sourceTagMap)
```

Comment:
I find it peculiar having 3 different metrics that capture similar things, with an introduction of a tag we can simplify things.

What if `gateway.write_key_requests` had two extra tags:
* `failed`: true/false 
* `reason` - same to what you introduce here.

Using the `failed` tag above you can get any metric you like.

With this, we need can simplify our code and make our stats more compact.

We still need to keep:

```go
		gateway.updateSourceStats(sourceSuccessStats, "gateway.write_key_successful_requests", sourceTagMap)
		gateway.updateFailedSourceStats(sourceFailStats, "gateway.write_key_failed_requests", sourceTagMap)
```

For a few release for backwards compatibility, but after that, they can be removed.

Reply:
@lvrach The similar metrics exist for gateway events, `gateway.write_key_events`, `gateway.write_key_successful_events`, `gateway.write_key_failed_events`

Let's take this a separate change ?

Also, I think we might need to revisit all the metrics once. I think we need to add rudderstack resourceIDs (`sourceID`, `destinationID`, `transformationID`, `trackingPlanID`) in the tags wherever possible. Going forward, this would enable us to send more notifications to the customers. With your changes to filter out these tags in hosted multi-tenant this should not be an issue ?

Reply:
+1 for tags